src/data_structures/kd_tree.rs

```diff
@@ -0,0 +1,470 @@
+use num_traits::{abs, real::Real, Signed};
+use std::iter::Sum;
+
+/// Internal node of a `KDTree`
+///
+/// `point` contains the data of the node
+/// `left` and `right` for branching
+struct KDNode<T: PartialOrd + Copy, const K: usize> {
+    point: [T; K],
+    left: Option<Box<KDNode<T, K>>>,
+    right: Option<Box<KDNode<T, K>>>,
+}
+
+impl<T: PartialOrd + Copy, const K: usize> KDNode<T, K> {
+    // Create a new node `KDNode` from a given point
+    fn new(point: [T; K]) -> Self {
+        KDNode {
+            point,
+            left: None,
+            right: None,
+        }
+    }
+}
+
+// A `KDTree` with a `root` node and a `size` to represent the number of points in the kd-tree
+pub struct KDTree<T: PartialOrd + Copy, const K: usize> {
+    root: Option<Box<KDNode<T, K>>>,
+    size: usize,
+}
+
+impl<T: PartialOrd + Copy, const K: usize> Default for KDTree<T, K> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<T: PartialOrd + Copy, const K: usize> KDTree<T, K> {
+    // Create and empty kd-tree
+    // #[must_use]
+    pub fn new() -> Self {
+        KDTree {
+            root: None,
+            size: 0,
+        }
+    }
+
+    // Returns true if point found, false otherwise
+    pub fn contains(&self, point: &[T; K]) -> bool {
+        search_rec(&self.root, point, 0)
+    }
```

Comment:
Is this really needed?

Reply:
`Contains` is useful to search for the presence of a point in a k-d tree.
Or do you mean the additional call of `search_rec`?

Reply:
Does `contains` need to be public?

Reply:
**Yes** for interaction with the outside world. However, `contains` is just to complete `CRUD` operations on a k-d tree.
Indeed, the current k-d tree implementation doesn't make use of `contains`. However, its implementation could help as it is just a synonym of the `search` method of a k-d tree.
Would you recommend I make it private for now or remove it?